### ISSUE TITLE
Problem reading mmCIF file

### ISSUE THREAD

**Issue by josemduarte** [dkoes/3Dmol.js#65]
I've read in some old issues that mmCIF support is implemented already into 3dmol.js. Is that right?

I wanted to try a very simple case with a file downloaded straight from the PDB (PDB id 1smt). This is what I do:

```html
<html>

<head>
 <script src="http://3Dmol.csb.pitt.edu/build/3Dmol-min.js"></script> 
</head>

<div style="height: 800px; width: 800px; position: relative;" class='viewer_3Dmoljs' data-href='1smt.cif'>
</div>

</html>

```

I only get an empty page and this in the js console:

```
Uncaught TypeError: Cannot read property 'id' of undefined (3Dmol-min.js:529)
```

I've also tried removing all headers from the mmCIF file and keeping only the _atom_site lines, but the error is the same. Also if I specify the file format as `data-type='cif'` the error is the same.

Any ideas what am I doing wrong? 

Thanks by the way for such a great piece of software :)

**Comment by dkoes** [comment]
The code was failing due to the lack of a struct_conn record.  I've committed a fix (6edcd5f949685303f7bc47d0449b5ee018e5fb66) and it should work now.

**Comment by josemduarte** [comment]
So should it be working for structures that do contain `_struct_conn` records?  I've tried with 2 other structures (2trx and 1b8g) both having `_struct_conn` and I get an error too. For instance for 2trx:

```
TypeError: mmCIF._struct_conn.ptnr2_label_alt_id is undefined (3Dmol.js:24572:16)

```
(This time I'm using the unminified file at http://3Dmol.csb.pitt.edu/build/3Dmol.js)

Looking at the mmCIF file, indeed `_struct_conn.ptnr2_label_alt_id` doesn't exist, instead it's called `_struct_conn.pdbx_ptnr2_label_alt_id`

Thanks for the help!

**Comment by jus1in** [comment]
Should be fixed by #66

**Comment by josemduarte** [comment]
Thanks for the fix, it looks great now! One related issue I'm seeing now is that selections don't see to be working. This is my code:

```html
<html>

<head>
<script src="http://3Dmol.csb.pitt.edu/build/3Dmol.js"></script>
</head>

<div style="height: 800px; width: 800px; position: relative;" class='viewer_3Dmoljs' 
 data-href='1smt.cif' 
 data-backgroundcolor='0xffffff' 
 data-select1='chain:A' data-style1='cartoon:color=0x00ff00'>
</div>

</html>

```

However the structure is still displayed in the default lines representation. The equivalent code with a pdb file works fine.

**Comment by dkoes** [comment]
That's because we weren't reading in the chain info.  Fixed now: 19c8ebf

**Comment by josemduarte** [comment]
Thanks again for the quick fix. One doubt: should I be seeing the changes already in `http://3Dmol.csb.pitt.edu/build/3Dmol.js` or you guys build that file only after every official release? 
(At the moment I still experiment the chain selection problem using the current `http://3Dmol.csb.pitt.edu/build/3Dmol.js` )

And talking about chain selections and mmCIF, would multi-character chain identifiers be allowed already in selections? 

Thanks!

**Comment by dkoes** [comment]
The 3dmol server doesn't get automatically updated, but the intention is that the build directory has a development copy (as opposed to the release directory).

I don't see why multichain identifiers wouldn't work (of course they won't for PDB files).

**Comment by josemduarte** [comment]
Another related issue, this time very weirdly only happening in Firefox (38.0) but not in Chrome (43.0.2357.125). I'm testing both in Linux (Ubuntu 15.04).

The issue happens with reading gzipped coordinate files. I've seen that 3dmol.js can read gzipped files directly, e.g. if I have a file `1smt.pdb.gz`, this works (both in Firefox and Chrome):
```html
<html>
<head>
<script src="http://3Dmol.csb.pitt.edu/build/3Dmol.js"></script>
</head>
<div style="height: 800px; width: 800px; position: relative;" class='viewer_3Dmoljs' 
 data-href='1smt.pdb' >
</div>
</html>

```
However if what I have is `1smt.cif.gz` and I try this:
```html
<html>
<head>
<script src="http://3Dmol.csb.pitt.edu/build/3Dmol.js"></script>
</head>
<div style="height: 800px; width: 800px; position: relative;" class='viewer_3Dmoljs' 
 data-href='1smt.cif' >
</div>
</html>

```
Then it works in Chrome without a problem, but in Firefox I get this error in the console:
```
10:56:42.666 TypeError: mmCIF._atom_site is undefined1 3Dmol.js:24638:24
```
Thanks again!

**Comment by dkoes** [comment]
3Dmol.js doesn't have native support for gzipped files.  Your webserver and/or browser is decompressing the file for you (which is why the behavior is browser dependent). You might want to inspect the headers that are being sent.  You may see "Content-Encoding: x-gzip".

I suspect this is at least partially a function of your server setup.  I can't get Chrome or Firefox to auto-decompress a gzipped file.

pako is a JavaScript library for gzip decompression.
https://github.com/nodeca/pako

**Comment by josemduarte** [comment]
Thanks for the answer. I thought there was some kind of gzip support already and that's what got me confused. Indeed as you say this must be a feature of the server I am using. It is weird in any case that the behaviour is browser and file-type dependent. I'll have a better look at it and try to find what's going on.

In any case thanks for adding the [new issue](https://github.com/dkoes/3Dmol.js/issues/70)
